runtime_patch.go

```diff
@@ -60,6 +60,99 @@ func updateMagicValue(file *ast.File, magicValue uint32) {
 	}
 }
 
+// updateEntryOffset adds xor encryption for funcInfo.entryoff
+func updateEntryOffset(file *ast.File, entryOffKey uint32) {
+	var nameOffField string
+	entryOffUpdated := false
+
+	extractNameOff := func(node ast.Node) bool {
+		indexExpr, ok := node.(*ast.IndexExpr)
+		if !ok {
+			return true
+		}
+		selExpr, ok := indexExpr.Index.(*ast.SelectorExpr)
+		if !ok {
+			return true
+		}
+		nameOffField = selExpr.Sel.Name
+		return false
+	}
+
+	updateEntryOff := func(node ast.Node) bool {
+		callExpr, ok := node.(*ast.CallExpr)
+		if !ok {
+			return true
+		}
+
+		textSelExpr, ok := callExpr.Fun.(*ast.SelectorExpr)
+		if !ok || textSelExpr.Sel.Name != "textAddr" {
+			return true
+		}
+
+		selExpr, ok := callExpr.Args[0].(*ast.SelectorExpr)
+		if !ok {
+			return true
+		}
+
+		callExpr.Args[0] = &ast.BinaryExpr{
+			X:  selExpr,
+			Op: token.XOR,
+			Y: &ast.ParenExpr{
+				X: &ast.BinaryExpr{
+					X: ah.CallExpr(ast.NewIdent("uint32"), &ast.SelectorExpr{
+						X:   selExpr.X,
+						Sel: ast.NewIdent(nameOffField),
+					}),
+					Op: token.MUL,
+					Y: &ast.BasicLit{
+						Kind:  token.INT,
+						Value: strconv.FormatUint(uint64(entryOffKey), 10),
+					},
+				},
+			},
+		}
+		entryOffUpdated = true
+		return false
+	}
+
+	var entryFunc *ast.FuncDecl
+	var cfuncnameFunc *ast.FuncDecl
+
+	for _, decl := range file.Decls {
+		funcDecl, ok := decl.(*ast.FuncDecl)
+		if !ok {
+			continue
+		}
+		switch funcDecl.Name.Name {
+		case "entry":
+			entryFunc = funcDecl
+		case "cfuncname":
+			cfuncnameFunc = funcDecl
```

Comment:
if you make the Inspect calls here, then you don't need the extra code to store the FuncDecl variables and make the calls later.

then you can make the `for` loop return from the entire function once it's inspected both functions correctly. If you reach the code after the `for` loop, then that means you missed either of the funcs, so you can panic.

Reply:
It is important to call `extractNameOff` before `updateEntryOff`. If `cfuncname` is below `entry`, it will break logic because current `nameOff` is not known (it changes between versions)

Reply:
Ah I see. I hadn't noticed that. Perhaps document it.